File: src/menu_controller.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright (C) 2011 ~ 2015 Deepin, Inc.
#               2011 ~ 2015 Wang YaoHua
#
# Author:     Wang YaoHua <mr.asianwang@gmail.com>
# Maintainer: Wang YaoHua <mr.asianwang@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os

from PyQt5.QtGui import QCursor
from PyQt5.QtCore import QObject, pyqtSignal, pyqtSlot

from deepin_menu.menu import Menu, CheckableMenuItem

from i18n import _
from constants import MAIN_DIR

MENU_ICONS_DIR = os.path.join(MAIN_DIR, "image", "menu_icons")
menu_icon_normal = lambda x: os.path.join(MENU_ICONS_DIR,"%s-symbolic-small-norml.svg" % x)
menu_icon_hover = lambda x: os.path.join(MENU_ICONS_DIR, "%s-symbolic-small-hover.svg" % x)
menu_icon_tuple = lambda x: (menu_icon_normal(x), menu_icon_hover(x))

save_sub_menu = [
    CheckableMenuItem("save:radio:_op_auto_save", _("Autosave")),
    CheckableMenuItem("save:radio:_op_save_to_desktop", _("Save to desktop")),
    CheckableMenuItem("save:radio:_op_copy_to_clipboard", _("Copy to clipboard")),
    CheckableMenuItem("save:radio:_op_save_as", _("Save to specified folder")),
    CheckableMenuItem("save:radio:_op_copy_and_save", _("Autosave and copy to clipboard")),
]

right_click_menu = [
    ("_rectangle", _("Rectangle tool"), menu_icon_tuple("rectangle-tool")),
    ("_ellipse", _("Ellipse tool"), menu_icon_tuple("ellipse-tool")),
    ("_arrow", _("Arrow tool"), menu_icon_tuple("arrow-tool")),
    ("_line", _("Brush tool"), menu_icon_tuple("line-tool")),
    ("_text", _("Text tool"), menu_icon_tuple("text-tool")),
    None,
    ("_save", _("Save"), menu_icon_tuple("save")),
    ("_share", _("Share"), menu_icon_tuple("share")),
    ("_exit", _("Exit"), menu_icon_tuple("exit")),
]

class MenuController(QObject):
    toolSelected = pyqtSignal(str, arguments=["toolName"])
    saveSelected = pyqtSignal(int, arguments=["saveOption"])
    shareSelected = pyqtSignal()
    exitSelected = pyqtSignal()

    preMenuShow = pyqtSignal()
    postMenuHide = pyqtSignal()

    def __init__(self):
        super(MenuController, self).__init__()

    def _menu_unregistered(self):
        self.postMenuHide.emit()

    def _menu_item_invoked(self, _id, _checked):
        self.postMenuHide.emit()

        if _id == "_rectangle":
            self.toolSelected.emit("_rectangle")
        if _id == "_ellipse":
            self.toolSelected.emit("_ellipse")
        if _id == "_arrow":
            self.toolSelected.emit("_arrow")
        if _id == "_line":
            self.toolSelected.emit("_line")
        if _id == "_text":
            self.toolSelected.emit("_text")

        if _id == "save:radio:_op_auto_save":
            self.saveSelected.emit(1)
        if _id == "save:radio:_op_save_to_desktop":
            self.saveSelected.emit(0)
        if _id == "save:radio:_op_copy_to_clipboard":
            self.saveSelected.emit(4)
        if _id == "save:radio:_op_save_as":
            self.saveSelected.emit(2)
        if _id == "save:radio:_op_copy_and_save":
            self.saveSelected.emit(3)

        if _id == "_share":
            self.shareSelected.emit()
        if _id == "_exit":
            self.exitSelected.emit()

    @pyqtSlot(int)
    def show_menu(self, saveOption):
        self.preMenuShow.emit()

        self.menu = Menu(right_click_menu)
        self.menu.getItemById("_save").setSubMenu(Menu(save_sub_menu))

        self.menu.getItemById("save:radio:_op_auto_save").checked = \
            saveOption == 1
        self.menu.getItemById("save:radio:_op_save_to_desktop").checked = \
            saveOption == 0
        self.menu.getItemById("save:radio:_op_copy_to_clipboard").checked = \
            saveOption == 4
        self.menu.getItemById("save:radio:_op_save_as").checked = \
            saveOption == 2
        self.menu.getItemById("save:radio:_op_copy_and_save").checked = \
            saveOption == 3

        self.menu.itemClicked.connect(self._menu_item_invoked)
        self.menu.menuDismissed.connect(self._menu_unregistered)
        self.menu.showRectMenu(QCursor.pos().x(), QCursor.pos().y())

File: controller/menucontroller.cpp
/*
 * Copyright (C) 2017 ~ 2017 Deepin Technology Co., Ltd.
 *
 * Maintainer: Peng Hui<penghui@deepin.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "menucontroller.h"
#include "utils/configsettings.h"

#include <QApplication>
#include <QStyleFactory>
#include <QDebug>

const QSize MENU_ICON_SIZE = QSize(14, 14);

MenuController::MenuController(QObject *parent)
    : QObject(parent)
    , m_ration(1)
{
    m_menu = new QMenu;
    m_menu->setFocusPolicy(Qt::StrongFocus);
    m_menu->setStyle(QStyleFactory::create("dlight"));

    QIcon rectIcon;
    rectIcon.addFile(":/image/menu_icons/rectangle-menu-norml.svg", MENU_ICON_SIZE,  QIcon::Normal);
    rectIcon.addFile(":/image/menu_icons/rectangle-menu-hover.svg", MENU_ICON_SIZE, QIcon::Active);
    QAction* rectAct = new QAction(rectIcon, tr("Rectangle"), this);
    connect(rectAct, &QAction::triggered, [=] {
        emit shapePressed("rectangle");
    });

    QIcon ovalIcon;
    ovalIcon.addFile(":/image/menu_icons/ellipse-menu-norml.svg", MENU_ICON_SIZE,  QIcon::Normal);
    ovalIcon.addFile(":/image/menu_icons/ellipse-menu-hover.svg", MENU_ICON_SIZE, QIcon::Active);
    QAction* ovalAct = new QAction(ovalIcon, tr("Ellipse"), this);
    connect(ovalAct, &QAction::triggered, [=]{
        emit shapePressed("oval");
    });

    QIcon arrowIcon;
    arrowIcon.addFile(":/image/menu_icons/arrow-menu-norml.svg", MENU_ICON_SIZE, QIcon::Normal);
    arrowIcon.addFile(":/image/menu_icons/arrow-menu-hover.svg", MENU_ICON_SIZE, QIcon::Active);
    QAction* arrowAct = new QAction(arrowIcon, tr("Arrow"), this);
    connect(arrowAct, &QAction::triggered, [=]{
        emit shapePressed("arrow");
    });

    QIcon penIcon;
    penIcon.addFile(":/image/menu_icons/line-menu-norml.svg", MENU_ICON_SIZE, QIcon::Normal);
    penIcon.addFile(":/image/menu_icons/line-menu-hover.svg", MENU_ICON_SIZE, QIcon::Active);
    QAction* penAct = new QAction(penIcon, tr("Pencil"), this);
    connect(penAct, &QAction::triggered, [=]{
        emit shapePressed("line");
    });

    QIcon textIcon;
    textIcon.addFile(":/image/menu_icons/text-menu-norml.svg", MENU_ICON_SIZE, QIcon::Normal);
    textIcon.addFile(":/image/menu_icons/text-menu-hover.svg", MENU_ICON_SIZE, QIcon::Active);
    QAction* textAct = new QAction(textIcon, tr("Text"), this);
    connect(textAct, &QAction::triggered, [=]{
        emit shapePressed("text");
    });

    QIcon unDoIcon;
    unDoIcon.addFile(":/image/menu_icons/undo-menu-normal.svg", MENU_ICON_SIZE, QIcon::Normal);
    unDoIcon.addFile(":/image/menu_icons/undo-menu-hover.svg", MENU_ICON_SIZE, QIcon::Active);
    QAction* unDoAct = new QAction(unDoIcon, tr("Undo"), this);
    connect(unDoAct, &QAction::triggered, [=]{
        emit unDoAction();
    });

    m_menu->addAction(rectAct);
    m_menu->addAction(ovalAct);
    m_menu->addAction(arrowAct);
    m_menu->addAction(penAct);
    m_menu->addAction(textAct);
    m_menu->addSeparator();
    m_menu->addAction(unDoAct);

    QIcon saveIcon;
    saveIcon.addFile(":/image/menu_icons/save-menu-norml.svg", MENU_ICON_SIZE, QIcon::Normal);
    saveIcon.addFile(":/image/menu_icons/save-menu-hover.svg", MENU_ICON_SIZE, QIcon::Active);
    QMenu* saveMenu =  m_menu->addMenu(saveIcon, tr("Save"));

    saveMenu->setStyle(QStyleFactory::create("dlight"));
    QAction* saveAct1 = new QAction(tr("Save to desktop"), this);
    QAction* saveAct2 = new QAction(tr("Autosave"), this);
    QAction* saveAct3 = new QAction(tr("Save to specified folder"), this);
    QAction* saveAct4 = new QAction(tr("Copy to clipboard"), this);
    QAction* saveAct5 = new QAction(tr("Autosave and copy to clipboard"), this);
    QList<QAction*> actionList;
    actionList.append(saveAct1);
    actionList.append(saveAct2);
    actionList.append(saveAct3);
    actionList.append(saveAct4);
    actionList.append(saveAct5);
    for(int k = 0; k < actionList.length(); k++) {
        saveMenu->addAction(actionList[k]);
        connect(actionList[k], &QAction::triggered, [=]{
            emit saveBtnPressed(k);
        });
    }

    int saveOptionIndex = ConfigSettings::instance()->value("save", "save_op").toInt();
    actionList[saveOptionIndex]->setCheckable(true);
    actionList[saveOptionIndex]->setChecked(true);

    QIcon exitIcon;
    exitIcon.addFile(":/image/menu_icons/exit-menu-norml.svg", MENU_ICON_SIZE, QIcon::Normal);
    exitIcon.addFile(":/image/menu_icons/exit-menu-hover.svg", MENU_ICON_SIZE, QIcon::Active);
    QAction* closeAct = new QAction(exitIcon, tr("Exit"), this);
    m_menu->addAction(closeAct);
    connect(closeAct, &QAction::triggered, this, [=]{
        emit shapePressed("close");
    });

    connect(m_menu, &QMenu::aboutToHide, this, [=]{
        emit menuNoFocus();
    });
}

void MenuController::showMenu(QPoint pos) {
    m_menu->popup(pos);
}

MenuController::~MenuController() {

}
